addons/crm/static/src/js/forecast/forecast_service.js

```diff
@@ -0,0 +1,230 @@
+         * 
+         * (3) compute the additional steps needed to reach the last position of (C), from the position reached in (2):
+         *     {cycle - (min_groups - 1) % cycle}
+         * 
+         * (4) combine (1) and (3), the sum should not be greater than a full cycle (-> truncate):
+         *     {(2 * cycle - (min_groups - 1) % cycle - cyclePos) % cycle}
+         * 
+         * (4) add min_groups!
+         */
+        var forecastPeriod = (2 * cycle - (this.min_groups - 1) % cycle - cyclePos) % cycle + this.min_groups;
+        this.end = moment(this.start).add(forecastPeriod, `${this.granularity}s`);
+    }
+    /**
+     * Adds one "granularity" period to [this.end], to expand the current forecast period
+     */
+    expand() {
+        this.end.add(1, `${this.granularity}s`);
+    }
+    /**
+     * @param {Array[]} domain
+     * @param {boolean} limited whether [this.end] moment must be used as a domain constraint to limit
+     *                          read_group results [limited === true]
+     *                          or not [limited === false]
+     * @returns {Array[]} modified domain
+     */
+    getDomain(domain, limited=undefined) {
+        if (typeof limited === 'undefined') {
+            limited = !!this.end;
+        }
+        if (!limited) {
+            return domain;
+        }
+        var extraDomain = ['|', [this.fieldName, '=', false],
+            [this.fieldName, '<', this.end.format(this.format)]];
+        if (domain.length) {
+            domain = ['&', ...domain, ...extraDomain];
+        } else {
+            domain = extraDomain;
+        }
+        return domain;
+    }
+    /**
+     * @param {Object} context
+     * @param {boolean} greedy whether fill_temporal must apply until [this.end] [greedy === true]
+     *                         or until the last group with at least one record [greedy === false]
+     * @returns {Object} modified context
+     */
+    getContext(context, greedy=undefined) {
+        if (typeof greedy === 'undefined') {
+            greedy = !!this.end;
+        }
+        context = {
+            ...context,
+            fill_temporal: {
+                fill_from: this.start.format(this.format),
+                min_groups: this.min_groups
+            }
+        };
+        if (greedy) {
+            context.fill_temporal.fill_to = moment(this.end).subtract(1, this.minGranularity).format(this.format);
+        }
+        return context;
+    }
+    /**
+     * @param {integer} min_groups minimum amount of groups to display, regardless of other constraints
+     */
+    set_min_groups(min_groups) {
+        this.min_groups = min_groups || 1;
+    }
+}
+
+/**
+ * Forecast Service
+ *
+ * This service will be used to generate or retrieve Forecast Periods
+ * 
+ * A specific forecast period configuration will always refer to the same instance
+ * unless reset is true
+ */
+ var ForecastService = AbstractService.extend({
```

Comment:
I quite like `FillTemporalService`, but as it is now, it only meets the requirements for my use case (Forecast on CRM leads), to really deserve this name, it would require an update to add more flexibility over fill_from, fill_to and the corresponding domain constraints. But we could update the name and update the code later when it's needed ^^

Reply:
As you prefer. ForecastService reads a bit generic, especially when grepping "Forecast" in your JS code :D .